### PR TITLE
refactor rdata

### PR DESCRIPTION
**Pull request by FFroehlich**
Goal is to completely remove any access to rdata in forward and backward problem. This will allow us to

- reduce memory impact of rdata since we don't need to prefill any results but can compute everything ad hoc (given a model and solver instance)
- implement hierarchical optimization in runAmiciSimulations with adjoints since we can seperately loop over forward problems, then compute noise parameters and then loop over backward problems.
- better debugging of integration failures as we can implement a more fine-grained access to the solution data in solvers after integration failure.

Both topics should and will be adressed in later PRs. May also split the rdata refactoring into multiple PRs.

### CODE REVIEW

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
bad news, `CVodeGetDky` only provides interpolated solutions in the current step size interval. Thus both `x` (and `sx` for FSA) will always have to be stored online. Shouldn't affect points 1) and 2), but makes 3) a little bit trickier.

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

File: src/edata.cpp [hunk 1]
```diff
@@ -389,13 +391,45 @@ void ConditionContext::applyCondition(const ExpData *edata)
...
+      if (!edata->fixedParametersPresimulation.empty()) {
+          // fixed parameter in model are superseded by those provided in
+          // edata
+          if (edata->fixedParametersPresimulation.size()
+              != (unsigned)model->nk())
+              throw AmiException("Number of fixed parameters (%d) in model does"
+                                 " not match ExpData (presimulation) (%zd).",
+                                 model->nk(),
+                                 edata->fixedParametersPresimulation.size());
+          model->setFixedParameters(edata->fixedParameters);
```
Comment (dweindl) [inline src/edata.cpp:430]:
```suggestion
          model->setFixedParameters(edata->fixedParametersPresimulation);
```
?

File: src/misc.cpp [hunk 2]
```diff
@@ -31,6 +31,29 @@ gsl::span<realtype> slice(std::vector<realtype> &data, const int index,
...
 }
 
+const gsl::span<const realtype> slice(const std::vector<realtype> &data,
+                                      const int index,
+                                      const unsigned size) {
+    if ((index + 1) * size > data.size())
+        throw std::out_of_range("requested slice is out of data range");
+    if (size > 0)
+        return gsl::make_span(&data.at(index*size), size);
+    else
```
Comment (dweindl) [inline src/misc.cpp:41]:
```suggestion

```

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/misc.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- src/misc.cpp  2
- src/model_ode.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/misc.cpp  2
- src/model_ode.cpp  2
- src/model.cpp  6
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- src/misc.cpp  2
- src/model.cpp  6
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- src/misc.cpp  2
- src/model_ode.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- src/model.cpp  6
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- include/amici/misc.h  2
- src/model_ode.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- src/model_ode.cpp  2
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
As a little guidance on what was done and why:

- split forwardproblem and pre/post equilibration to be run seperately
- remove any notion of rdata from {Forward,SteadyState,Backward}Problem
- introduce storage variables and add them to problems to facilitate post-hoc computation of rdata

**Comment by FFroehlich** [comment]
test_steadystate currently fails since jacobian&xdot are now stored in rdata upon integration failure. will update as soon as everything else is resolved.

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- include/amici/misc.h  2
- src/model_ode.cpp  2
- src/model.cpp  6
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- src/model.cpp  6
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- include/amici/misc.h  2
- src/model.cpp  6
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- src/model_ode.cpp  2
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- include/amici/misc.h  2
- src/model.cpp  6
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- src/model_ode.cpp  2
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pysb.py  4
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- include/amici/misc.h  2
- src/model.cpp  6
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pysb.py  4
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- src/model_ode.cpp  2
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pysb.py  4
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- include/amici/misc.h  2
- src/model.cpp  6
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

File: src/model_dae.cpp [hunk 3]
```diff
@@ -110,22 +111,24 @@ void Model_DAE::fdxdotdp(const realtype t, const N_Vector x,
         throw AmiException("Wrapping of DAEs is not yet implemented from Python");
     } else {
         // matlab generated
-        fdwdp(t, N_VGetArrayPointer(x_pos)); // Why is it x_pos here and x ind model_ode.cpp?
+        fdwdp(t, N_VGetArrayPointer(x_pos));
+        // Why is it x_pos here and x ind model_ode.cpp?
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply src/model_dae.cpp:115]: its not :D

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pysb.py  4
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model.cpp  6
- src/model_ode.cpp  2
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pysb.py  4
- python/tests/test_preequilibration.py  1
         

Clones added
============
- src/model_ode.cpp  2
- src/model.cpp  6
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pysb.py  4
- python/tests/test_preequilibration.py  1
         

Clones added
============
- include/amici/misc.h  2
- src/model.cpp  6
- src/model_ode.cpp  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_preequilibration.py  1
- python/tests/test_pysb.py  4
         

Clones added
============
- src/model_ode.cpp  2
- src/model.cpp  6
- include/amici/misc.h  2
         

Clones removed
==============
+ src/forwardproblem.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5147305&bid=16907353)

**Pull request merged**